### ISSUE TITLE
Revisit the facet umap function

### ISSUE THREAD

**Issue by sjspielman** [AlexsLemonade/scpca-nf#637]
See original comment: https://github.com/AlexsLemonade/scpca-nf/pull/635#issuecomment-1868031802

The legend placement is not sufficiently general for the plots we'll be using it for, so we'll need to revisit this code.

- We don't lump cell types for submitters, so the legend needs to be placed "normally" outside the margins
- For singler and cellassign, the positioning may also be dependent on text length

**Comment by sjspielman** [comment]
We should also figure out if we really do or do not want `theme(aspect.ratio = 1)` in these plots.

**Comment by jashapiro** [comment]
In my opinion, we probably should lump submitter annotations for these plots too.

For the text length problem, we may be able to solve it with modifying legend placement to something like:

```
legend.position = c(.66, .33),
legend.justification = c("left", "top"),
```

**Comment by jashapiro** [comment]
> We should also figure out if we really do or do not want `theme(aspect.ratio = 1)` in these plots.

I think we do, in general, but it will require that we get the dimensions of the plot as a whole to match: if we do 3X3, we need the plot to be pretty square. If we have more than the 8 plots, then we will need a matching aspect ratio (and to deal with the legend differently). 

One other option is to move some of the theme settings back out of the function that we are using... We can always add them back on a per-plot basis.

```
faceted_umap(...) + theme(...)
```

**Comment by sjspielman** [comment]
> In my opinion, we probably should lump submitter annotations for these plots too (move the lumping code into this function?).

Noting that I was wrong - we _do_ already lump submitter cell types; it's just that my brain was toast the day before break 😬 . The code lumps/mutates at `across(ends_with("_celltype_annotation"))`, which includes all the annotation columns (singler, cellassign, and submitter).

So, we'll want to count the final number of cell types and determine appropriate layout from there to achieve some kind of 3x3 grid. I might vote something like this for N<=8, where dashed-line boxes are panels that may or may not be there depending on the actual N, and the lines with pink dots are legends. The yellow text is a loose concept of where to place the legend - 

<img src="https://github.com/AlexsLemonade/scpca-nf/assets/4701111/9386d257-390e-4acc-8288-104b447156a8" width="300">

Extrapolating, we can use `%%3` to conditionally handle legend placement:
- if `N == 1 | N%%3 == 0`, legend goes on the bottom
  - Or, for `N==1`, unlike my drawing, we can put the legend on the right
- If `N %%3 == 2`, legend goes at 0.66
- If `N %%3 == 1`, legend goes at 0.33 

The only time things would get out of handle with a 3-column layout is if there are a _lot of clusters._ Perhaps we cap the situation at like N=15, and change the ncol for anything bigger than that?

I'll note also I previously brought up the idea that we might not _need_ a legend at all since the facets are all labeled anyways. This was nixed at the time, but maybe it's more of a contender now?


Note I also intend to use `str_wrap()` to play around with overall label sizes.

Tagging @allyhawkins @jashapiro for post-winter-break thoughts!

**Comment by allyhawkins** [comment]
> * if `N == 1 | N%%3 == 0`, legend goes on the bottom
>   
>   * Or, for `N==1`, unlike my drawing, we can put the legend on the right
> * If `N %%3 == 2`, legend goes at 0.66
> * If `N %%3 == 1`, legend goes at 0.33

This approach seems reasonable to me. 

> The only time things would get out of handle with a 3-column layout is if there are a _lot of clusters._ Perhaps we cap the situation at like N=15, and change the ncol for anything bigger than that?

I think we had decided that we would use N = 7 for lumping? That's our default and we don't actually change it in the workflow so I would not anticipate having any more than 7 cell types +  1 for all other cell types in any of the UMAPs. If you really wanted to be careful we could add a check to the `lump_celltypes` function to make sure N<=7. So I really think we just need to consider the situations where we have less than 8 categories. I wouldn't worry about the situations where we have more since that won't happen. 


> I'll note also I previously brought up the idea that we might not _need_ a legend at all since the facets are all labeled anyways. This was nixed at the time, but maybe it's more of a contender now?

I also would be okay with removing the legend. The reason I do like having it is because the colors are different for each cell type so then users would know what color is expected in that panel.

**Comment by sjspielman** [comment]
Yes - n=7 for lumping but then the extra "all other cell types" brings us to 8 total. I think we're saying the same thing?

**Comment by allyhawkins** [comment]
Yes, we are. But I don't know why we would ever have a scenario with N=15.

**Comment by sjspielman** [comment]
> But I don't know why we would ever have a scenario with N=15.

It definitely wouldn't happen with cell types, but possibly with clusters.

**Comment by sjspielman** [comment]
@jashapiro and I had a quick chat about this on Slack. Notes - 

- We'd probably like to avoid excessive logic since it's preferable to keep code changes more minimal
- Let's make our lives easier:
  - If 7 or 8 categories, legend goes in the bottom right at ~0.66 (depending on fiddling). We expect this to be the vast majority situations anyways. Cell type methods indeed enjoy assigning cell types. 
  - Anything else, just put the legend on the bottom.

**Comment by sjspielman** [comment]
Closed by #644
